### PR TITLE
Pytorch Lightning Integration

### LINKED ISSUE #568: Integrate PyTorch Lightning
I've been pretty skeptical about high level PyTorch libraries to reduce boilerplate, but PyTorch Lightning seems to be the real deal and a good long term solution for support! (more info [here](https://github.com/PyTorchLightning/pytorch-lightning)).

Integration will include a bunch of fixes I'd like to make as well, which will break backwards compatibility. This means a new model iteration will need to be done but it's overdue.

To make sure I don't miss anything here's a checklist:

- [x] Feature parity in training/testing/loading models
- [x] Save/Load models during training using the same exposed checkpointing parameters
- [x] Ensure GCS based checkpointing is included
- [x] Add Trains/Visdom custom logging module (seems there is already integration for Trains [here](https://github.com/PyTorchLightning/pytorch-lightning-bolts) after speaking to @williamFalcon)
- [x] Update README/Instructions for scaling, validate scaling parameters including data samplers etc

Final Steps:

- [ ] Final model iteration across all datasets, and potentially a combined dataset if we can find someone with resources!

### PR DESCRIPTION
**Pull request by SeanNaren**
Closes #568

### CODE REVIEW

**Comment by SeanNaren** [comment]
Next steps are to integrate saving/loading of the model from training to eval.

File: README.md [hunk 1]
```diff
@@ -26,20 +26,14 @@ an Anaconda installation on Ubuntu, with PyTorch installed.
 
 Install [PyTorch](https://github.com/pytorch/pytorch#installation) if you haven't already.
 
-Install this fork for Warp-CTC bindings:
+Install Warp-CTC bindings:
```
Comment (igormq) [inline README.md:29]: Is Warp-CTC still needed? Have you already tried use the native Pytorch CTC? They have fixed several issues regarding the loss (see [#27460](https://github.com/pytorch/pytorch/pull/27460), [#27039](https://github.com/pytorch/pytorch/pull/27039), [#23298](https://github.com/pytorch/pytorch/pull/23298)).  I know that are some issues, but the warp ctc has issues too. I have trained a couple of months ago with the pytorch ctc and I was able to get pretty decent results :+1:

Comment (SeanNaren) [reply README.md:29]: Thanks @igormq! I agree, there are still issues with the native CTC loss but it makes support much much easier. I'll do a few test runs locally to see how it stacks up, and if it seems reasonable I'll replace warp-ctc with the native function.

I'm just still hesitant as there are edge cases that are unexplained still in the native CTC function. Atleast warp-ctc I've tried and tested it in many scenarios. 

Can't wait to get rid of warp-ctc 😂

File: deepspeech_pytorch/model.py [hunk 2]
```diff
@@ -156,34 +163,53 @@ def __init__(self, rnn_type, labels, rnn_hidden_size, nb_layers, audio_conf,
...
-            nn.BatchNorm1d(rnn_hidden_size),
-            nn.Linear(rnn_hidden_size, num_classes, bias=False)
+            nn.BatchNorm1d(self.model_cfg.hidden_size),
+            nn.Linear(self.model_cfg.hidden_size, num_classes, bias=False)
         )
         self.fc = nn.Sequential(
             SequenceWise(fully_connected),
         )
         self.inference_softmax = InferenceBatchSoftmax()
+        self.criterion = CTCLoss(reduction='sum', zero_infinity=True)
```
Comment (piraka9011) [inline deepspeech_pytorch/model.py:203]: I'm currently using `reduction='mean'` and `zero_infinity=False` (the defaults)
What's the reason for these parameters?

File: deepspeech_pytorch/data/utils.py [hunk 3]
```diff
@@ -1,38 +1,65 @@
...
+        'root_path': data_path,
+        'samples': []
+    }
+    for wav_path in tqdm(file_paths, total=len(file_paths)):
+        wav_path = str(wav_path.relative_to(data_path))
+        transcript_path = wav_path.replace('/wav/', '/txt/').replace('.wav', '.txt')
+        manifest['samples'].append({
+            'wav_path': wav_path,
+            'transcript_path': transcript_path
+        })
```
Comment (piraka9011) [inline deepspeech_pytorch/data/utils.py:42]: I've found a more robust way to do this was the following.

```suggestion
        wav_path = wav_path.relative_to(data_path)
        transcript_path = wav_path.parent.with_name("txt") / wav_path.with_suffix(".txt").name
        manifest['samples'].append({
            'wav_path': wav_path.as_posix(),
            'transcript_path': transcript_path.as_posix()
        })
```

Otherwise, the string replacement might not work with files with certain names.

File: deepspeech_pytorch/model.py (hunk 2 above)
Comment (SeanNaren) [reply deepspeech_pytorch/model.py:203]: Reduction sum means do not average the loss/gradients. This is reflecting the behaviour of warp-ctc. Zero infinity means if we get an inf gradient, we zero instead of returning infinite. Another behaviour copying warp-ctc!

File: deepspeech_pytorch/data/utils.py (hunk 3 above)
Comment (SeanNaren) [reply deepspeech_pytorch/data/utils.py:42]: Epic! Thanks so much

**Comment by SeanNaren** [comment]
I've run into some strange convergence issues with librispeech on this branch. Going to investigate and have a look (might just be the batch size needs to be small).

**Comment by piraka9011** [comment]
FYI, I had to install `omegaconf` from source (specifically from [this](https://github.com/omry/omegaconf/commit/a00ea90052c1fcdcda9cf86853737ec3cd213f5c) commit) in order to get `DeepSpeech.load_from_checkpoint()` to work, otherwise, I would get an error when the model tried to parse the `spect_cfg.sample_rate` from the checkpoint.

Maybe this is an upstream issue w/ PTL?

```py
TypeError: issubclass() arg 1 must be a class
```

File: deepspeech_pytorch/inference.py [hunk 4]
```diff
@@ -69,13 +80,12 @@ def run_transcribe(audio_path: str,
...
-                   use_half: bool):
+                   precision: int):
     spect = spect_parser.parse_audio(audio_path).contiguous()
     spect = spect.view(1, 1, spect.size(0), spect.size(1))
     spect = spect.to(device)
-    if use_half:
-        spect = spect.half()
     input_sizes = torch.IntTensor([spect.size(3)]).int()
-    out, output_sizes = model(spect, input_sizes)
+    with autocast(enabled=precision is 16):
```
Comment (piraka9011) [inline deepspeech_pytorch/inference.py:88]: This will result in a `SyntaxWarning`. Ditto for `model.py` and `validation.py`

Suggested way was
```suggestion
    with autocast(enabled=precision == 16):
```

**Comment by SeanNaren** [comment]
Builds will fail due to lack of pre-trained models, but that should be solved once i make the release! I think we're there

**Comment by ritwikmishra** [comment]
> ```python
> issubclass
> ```

I solved this by reinstalling omegaconf to this version ```omegaconf==2.1.0.dev9```.
